task/control-plane-agent/src/mgs_sidecar.rs

```diff
@@ -166,19 +237,204 @@ impl MgsHandler {
+        if let Err(e) = net.trust_vlan(vid, trust_until) {
+            ringbuf_entry!(Trace::TrustVLanFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+        ringbuf_entry!(Trace::UnlockedUntil(vid, trust_until));
+
+        // Reconfigure our own internal state to accept messages
+        self.locked[vlan_index] = LockState::UnlockedUntil(trust_until);
+
+        Ok(())
+    }
+
+    fn lock(&mut self, vid: u16) -> Result<(), GwMonorailError> {
+        ringbuf_entry!(Trace::Locking(vid));
+        let i = VLANS
+            .iter()
+            .position(|v| v.vid == vid)
+            .ok_or(GwMonorailError::InvalidVLAN)?;
+        self.locked[i] = LockState::Locked;
+
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        net.distrust_vlan(vid).map_err(|e| {
+            ringbuf_entry!(Trace::DistrustVLanFailed(e));
+            GwMonorailError::UnlockFailed
+        })?;
+        self.monorail.lock_vlans().map_err(|e| {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            GwMonorailError::UnlockFailed
+        })?;
+
+        Ok(())
+    }
+
+    fn is_sender_trusted(
+        &mut self,
+        sender: Sender,
+    ) -> Result<(), GwMonorailError> {
+        use gateway_messages::SpPort as GwSpPort;
+        use task_net_api::SpPort as NetSpPort;
+
+        // If this message is arriving on a trusted VLAN, then the lock state is
+        // irrelevant.
+        let i = task_net_api::VLANS
+            .iter()
+            .position(|v| {
+                matches!(
+                    (v.port, sender.port),
+                    (NetSpPort::One, GwSpPort::One)
+                        | (NetSpPort::Two, GwSpPort::Two)
+                ) && v.vid == sender.vid
+            })
+            .ok_or(GwMonorailError::InvalidVLAN)?;
+        if VLANS[i].trusted {
+            ringbuf_entry!(Trace::MessageTrusted(VLANS[i].vid));
+            return Ok(());
+        }
+
+        if let LockState::UnlockedUntil(t) = self.locked[i] {
+            let now = sys_get_timer().now;
+            if now < t {
+                ringbuf_entry!(Trace::MessageTrusted(VLANS[i].vid));
+                Ok(())
+            } else {
+                ringbuf_entry!(Trace::TimedRelock(VLANS[i].vid));
+                self.lock(VLANS[i].vid)?;
+                ringbuf_entry!(Trace::MessageNotTrusted(VLANS[i].vid));
+                Err(GwMonorailError::ManagementNetworkLocked)
+            }
+        } else {
+            ringbuf_entry!(Trace::MessageNotTrusted(VLANS[i].vid));
+            Err(GwMonorailError::ManagementNetworkLocked)
+        }
+    }
 }
 
 impl SpHandler for MgsHandler {
     type BulkIgnitionStateIter = ignition_handler::BulkIgnitionStateIter;
     type BulkIgnitionLinkEventsIter =
         ignition_handler::BulkIgnitionLinkEventsIter;
 
+    /// Checks whether we trust the given message
+    fn is_request_trusted(
```

Comment:
Having a function called `is_condition` that doesn't return a `bool` is weird. Maybe `ensure_request_trusted`?

This formulation also has "parse don't validate" implications. It would be neat if a trusted message had a different type via newtype or something; this would clearly distinguish which points in the code need to have a trusted message vs not. Not mandatory, but something to think about.